### PR TITLE
chore: resetGlobal fix and makeParentsDependOnChild add dependency optimisation

### LINKED ISSUE #32811: resetGlobalScope and updateDependencyMap addNode optimisation
This as part of eval v4 roadmap project#2 optimisation, over here we optimisate resetGlobalScope and optimise the addNode operation in the updateDependencyMap function. Limiting the addNode to only affected nodes in the dependencyMap will be handled in a separate PR.

### PR DESCRIPTION
**Pull request by vsvamsi1**
## Description
This as part of eval v4 roadmap project#2 optimisation, over here we optimisate resetGlobalScope and the addNode operation in the updateDependencyMap function. The following are the optimisations:

- resetWorkerGlobalScope changed library and keyword lookup from array to set, this is to reduce the time complexity.
- triggering resetWorkerGlobalScope after the first return statement, to remove redundant calls when code returns early.
- in makeParentsDependOnChild limiting the add operations getting triggered when the dependencies are the same.
- Added a library accessor code to help in generating a set when required.

 Noticed a 15% reduction in webworker scripting for a customer's app. Limiting the addNode to only affected nodes in the dependencyMap will be handled in a separate PR.

Fixes #32811  
> [!WARNING]  
> _If no issue exists, please create an issue first, and check with the maintainers if the issue is valid._

## Automation

/ok-to-test tags="@tag.All"

### :mag: Cypress test results

> [!TIP]
> 🟢 🟢 🟢 All cypress tests have passed! 🎉 🎉 🎉
> Workflow run: <https://github.com/appsmithorg/appsmith/actions/runs/8782269083>
> Commit: bcc890545851d9ba76beef4036fda98b9d50bd0e
> Cypress dashboard url: <a href="https://internal.appsmith.com/app/cypress-dashboard/rundetails-65890b3c81d7400d08fa9ee5?branch=master&workflowId=8782269083&attempt=2" target="_blank">Click here!</a>

## Communication
Should the DevRel and Marketing teams inform users about this change?
- [ ] Yes
- [ ] No

## Summary by CodeRabbit

- **New Features**
	- Introduced new testing for dependency handling in `DependencyMapUtils`.
	- Added a new Cypress test spec for ClientSide SettingsPane.

- **Refactor**
	- Enhanced dependency management in `DependencyMapUtils` to prevent redundant dependencies.
	- Updated handling of JavaScript libraries in the evaluation workflow to improve performance and accuracy.

- **Tests**
	- Expanded test coverage to ensure robust dependency management and JavaScript library handling.

- **Documentation**
	- Updated existing summaries to reflect changes in JavaScript library management and test specifications.

### CODE REVIEW

File: app/client/src/entities/DependencyMap/DependencyMapUtils.ts [hunk 1]
```diff
@@ -61,9 +60,15 @@ export class DependencyMapUtils {
...
-      const newDeps = union(existingImmediateParentDeps, [curKey]);
-
-      dependencyMap.addDependency(immediateParent, newDeps);
+      const existingImmediateParentDepsSet = new Set(
+        existingImmediateParentDeps,
+      );
+      // Add child to immediate parent's dependencies if not already present
+      // don't perform addDependency unnecessarily
+      if (!existingImmediateParentDepsSet.has(curKey)) {
+        const newDeps = [...existingImmediateParentDeps, curKey];
```
Comment (rajatagrawal) [inline app/client/src/entities/DependencyMap/DependencyMapUtils.ts:69]: Would it be possible to append curKey to existingImmediateParentDepsSet intead of creating a new copy of it ?

File: app/client/src/entities/DependencyMap/__tests__/index.test.ts [hunk 2]
```diff
@@ -2708,4 +2708,41 @@ describe("Tests for DependencyMapUtils", () => {
...
+    test("should trigger addDependency when the child node isn't there ", () => {
+      const spy = jest.spyOn(dependencyMap, "addDependency");
+      DependencyMapUtils.makeParentsDependOnChild(
+        dependencyMap,
+        "apiData.data",
+      );
+      expect(spy).toHaveBeenCalledTimes(1);
+    });
+    test("should not trigger addDependency when the child node is there ", () => {
+      //when apiData.data child node is already there
```
Comment (rajatagrawal) [inline app/client/src/entities/DependencyMap/__tests__/index.test.ts:2736]: It seems like this set is dependent on the execution of the first test. Would it be possible to make the execution of both the tests independent of each other ?

File: app/client/src/entities/DependencyMap/DependencyMapUtils.ts (hunk 1 above)
Comment (vsvamsi1) [reply app/client/src/entities/DependencyMap/DependencyMapUtils.ts:69]: Sure

File: app/client/src/workers/common/JSLibrary/index.ts [hunk 3]
```diff
@@ -32,6 +32,27 @@ export const defaultLibraries: JSLibrary[] = [
 ];
 
 export const JSLibraries = [...defaultLibraries];
+
+const JSLibraryAccessorModifier = () => {
+  let jsLibraryAccessorSet = Object.freeze(
+    new Set(JSLibraries.flatMap((lib) => lib.accessor)),
+  );
```
[the remark this answers is not in the document]
Comment (vsvamsi1) [reply app/client/src/workers/common/JSLibrary/index.ts]: Yeah its not necessary, a Set cannot be frozen.

**Review by rajatagrawal — APPROVED** [review]

**Pull request merged**